mirgecom/boundary.py

```diff
@@ -135,7 +159,40 @@ def boundary_pair(self, discr, q, btag, **kwargs):
         # Form the external boundary solution with the new momentum
         bndry_soln = join_conserved(dim=dim, mass=int_cv.mass,
                                     energy=int_cv.energy,
-                                    momentum=ext_mom,
-                                    species_mass=int_cv.species_mass)
+                                    momentum=ext_mom)
 
-        return TracePair(btag, interior=int_soln, exterior=bndry_soln)
+        return bndry_soln
+
+    def av(self, discr, q, btag, **kwargs):
+        """Get the exterior solution on the boundary."""
+        # Grab some boundary-relevant data
+        dim = discr.dim
+        cv = split_conserved(dim, q)
+        actx = cv.mass[0].array_context
+
+        # Grab a unit normal to the boundary
+        normal = thaw(actx, discr.normal(btag))
+
+        # Get the interior soln
+        int_soln = discr.project("vol", btag, q)
+        bndry_q = split_conserved(dim, int_soln)
+
+        # create result array to fill
+        result = np.zeros(2+dim, dtype=object)
```

Comment:
The biggest problem I have with this is the assumption regarding the necessary size of the data structure.  That this line is needed at all is already suspicious, and that it is a regular *numpy* array filled with zeros is again also odd to me. 

I feel like this whole thing could be replaced by something like (not exact here, I'm sure):

```
        bnd_cv = split_conserved(dim, int_soln)
        bnd_mom_normcomp = make_obj_array(
            [np.outer(normal, np.dot(bnd_cv.momentum, normal))[i]
            for i in range(dim)]
        )
        new_bnd_mom = 2*bnd_mom_normcomp - bnd_cv.momentum
        return join_conserved(dim=dim, mass=-bnd_cv.mass, energy=-bnd_cv.energy,
                              momentum=new_bnd_mom)
```

This would still need to be tweaked for multi-component gases, but at least it would help address some of the concerning things that stick out.

Reply:
I don't think join conserved here will work for how I currently have the arrays structured (unless it has relatively recently changed). There is definitely a discussion here to have about if I should be storing my data in a different structure. Let me try this and confirm it won't work first though.

Reply:
If it fails, it probably makes sense to track down why, and resolve that.  We should try to avoid creating and packing *numpy* arrays as it will likely result somewhere in an `object array <--> device` type issue.

Reply:
So, the suggested solution doesn't work. However, this does need to get fixed somehow so I can use join_conserved when using species arrays as well I suspect.

The root of the issue I believe to be is (this is me thinking out loud)

https://github.com/w-hagen/mirgecom/blob/cb8be6a102cb0b7089e724324761b97546480db9/mirgecom/artificial_viscosity.py#L131

discr.weak_grad() expects a DOFArray and returns and object array of size dim. Then when obj_array_vectorize is called with it, the result is a nested set of object arrays--an object array of size nvars, each containing an object array of size dim of DOFArrays. This structure is then not compatible with the join_conserved call.

It might make sense then to reshape this set of nested object arrays into a 2D object array? I think that would solve the issues here.

Reply:
Note to self: Promoting the nested objects to an array of (nvars,ndim) causes issues when calling `grudge.eager.cross_rank_trace_pairs`

Reply:
An attempt to clean this up can be found here: https://github.com/w-hagen/mirgecom/pull/11

This is also a bit strange with the functions needed to converter from nested objects to a 2d object and back. Maybe there is another way to handle this, but not sure I have any more ideas at the moment.

Reply:
Of potential interest: inducer/grudge#70